### PR TITLE
[JIT] Allow registering Decompositions 

### PR DESCRIPTION
**Pull request by eellison**
- Allow registering custom decompositions
- Add easier API for invoking decompositions
- Shorten API names (no users yet)

I am doing these as one pr because they are fairly short/simple and because github first does not support ghstack yet.

cc @Chillee @zou3519

### CODE REVIEW

**Review by davidberard98 — APPROVED** [review]
LGTM! had a small comment (also, fix the linter errors)

File: torch/csrc/jit/runtime/decomposition_registry.cpp [hunk 1]
```diff
@@ -21,6 +22,10 @@ auto compilation_unit = std::make_shared<CompilationUnit>();
 std::unordered_map<const FunctionSchema*, std::shared_ptr<Graph>>
     schema_to_decomposition;
 
+// Holds User-Registered Functions and keeps them alive
+std::unordered_map<const FunctionSchema*, std::shared_ptr<Function>>
+    user_registered_funcs;
+
```
[the remark this answers is not in the document]
Comment (eellison) [reply torch/csrc/jit/runtime/decomposition_registry.cpp:28]: The graph is kept alive, here we're keeping alive the `Function` so we have an executor for it

[the remark this answers is not in the document]
Comment (eellison) [reply torch/csrc/jit/runtime/decomposition_registry.cpp:28]: but youre right the comment isnt very good , will updat

**Comment by zengk95** [comment]
@pytorchbot revert this

This is breaking `linux-xenial-cuda11.3-py3.7-gcc7-bazel-test / build-and-test`. This was red on the PR as well. 
See this for logs https://hud.pytorch.org/pytorch/pytorch/commit/d9f0774f983007e32caed465ab83cc016deb3a9d

**Pull request closed without merging**